File: runtime/ruby/units/table/tail_ref.rb
# requires: rt/trap
# `call`'s checks, but returning the slot's tail-entry callable for the
# trampoline (ADR-18) instead of invoking: a slot's optional third element
# is the body method of a tail-calling function; entries without one
# complete in a single frame anyway.
def tail_ref(i, type_key)
  Rt.trap("undefined element") if i >= @slots.size
  slot = @slots[i]
  Rt.trap("uninitialized element") if slot.nil?
  ty, func, tail = slot
  Rt.trap("indirect call type mismatch") unless ty == type_key
  tail || func
end

File: crates/dewasm-backend-perl/units/table/tail_ref.pl
# requires: rt/trap
# `call`'s checks, raised at the same point in execution order, but returning the slot's coderef for the trampoline instead of invoking it.
# A slot's optional third element is a coderef for the body of a tail-calling function; a slot without one completes in a single frame anyway.
sub tail_ref {
    my ($self, $i, $type_key) = @_;
    Rt::trap('undefined element') if $i >= scalar @{$self->{slots}};
    my $slot = $self->{slots}[$i];
    Rt::trap('uninitialized element') unless defined $slot;
    Rt::trap('indirect call type mismatch') unless $slot->[0] eq $type_key;
    return $slot->[2] // $slot->[1];
}
